### ISSUE TITLE
ELI5 on these formats

### ISSUE THREAD

**Issue by DonaldTsang** [cdwfs/img2ktx#8]
- What are BC1, BC3, BC7, ETC1, and ASTC?
- What is KTX?
- Is this different for FLIF, FUIF or "PIK by Google"?

**Comment by cdwfs** [comment]
Sorry, didn't see this issue pop up until just now!

This tool is for converting image files into GPU-native block-compressed formats preferred by real-time graphics applications. Uncompressed image data can take up a lot of memory, which increases hardware requirements to run the application. Also, uncompressed image data results in worse texture cache utilization, and lower application performance. So, image compression is good!

Alas, common image file formats like PNG and JPEG are not supported as texture formats by GPUs, which require efficient access to individual pixels (these formats are optimized for decompressing an entire texture at once). Block-compressed formats offer a reasonable compromise; they're generally larger than PNG/JPEG/etc, but still offer roughly 4-8x compression over raw image data. They generally work by compressing small blocks of pixels independently (4x4 is a common block size) in such a way that locating and decompressing the block containing a specific pixel is very efficient on GPUs.

Now, to your actual questions:

> What are BC1, BC3, BC7, ETC1, and ASTC?

These are all different block-compressed texture formats. BCn are used by desktop GPUs on Windows/Mac/Linux/etc. ETC1 is, I believe, used on GPUs in older iOS devices. ASTC is a relatively new format, supported by Android GPUs (and possibly iOS as well at this point; I haven't been paying close enough attention).

>  What is KTX?

KTX is a container format for image files, specifically designed for storing texture data. If you're familiar with DDS files, it's designed to solve the same general problem in a less DirectX-specific way. Some advantages of storing texture files as KTX instead of individual PNG/JPEG files:
- KTX files can store data in GPU-native block-compressed formats. Compressing raw image data to these formats can be quite expensive, so it's usually desirable to compress offline at build time rather than every time the texture is loaded. They end up larger on disk as a side effect, but disk is cheap :)
- KTX files can store an entire mipmap chain in a single file. Mipmaps are lower resolutions of a given texture image, used by the GPU to improve visual quality. A 1024x1024 texture would have 9 additional mip levels (512x512, 256x256, etc. all the way down to 1x1), and storing them all in the same file is more convenient than loading them individually.
- KTX files can store all the faces of a cubemap in a single file. Cubemaps are a way of stitching together six images into a cube, such that the final image seems to surround the viewer at the center of the cube. Again, storing all six faces of a cubemap in a single file is more convenient than loading six individual files (especially since each face can have its own mipmap chain -- the number of individual source images can easily get way out of hand).

 > Is this different for FLIF, FUIF or "PIK by Google"?

I'm not familiar with any of these formats, but after a quick search: yes, this tool (and the formats it supports) is trying to solve a different problem than FLIF/FUIF/PIK (which seem targeted towards providing smaller on-disk whole-file compression for web applications).

Hope this helps!
